### PR TITLE
Feature/issue 92

### LINKED ISSUE #92: Add unit test infrastructure with WebSocket mocking
## Summary

All 107 existing tests are integration tests that require a live IBM i server connection. There are **zero unit tests** and **zero mocks**. This means:

- Tests cannot run in standard CI/CD without server access
- Tests cannot run offline during local development
- Error recovery paths and edge cases are untestable
- The test suite is slow (network I/O for every test)
- 7 pool tests are commented out due to job status tracking issues that could be debugged with mocks

## Current State

- **107 active tests**, all integration (require `VITE_SERVER`, `VITE_DB_USER`, `VITE_DB_PASS`)
- **No `conftest.py`** — no shared fixtures
- **No mocking** of WebSocket connections, server responses, or authentication
- **Empty test subdirectories** (`tests/unit/`, `tests/integration/`, etc.) — a planned reorg that was never completed
- **7 commented-out tests** in `pooling_test.py` and `pep249_test.py` due to job status tracking issues
- **pytest-cov installed but not used** — no coverage metrics collected
- **Estimated coverage**: ~55-65% overall, ~35% for error handling paths

### 1. Create shared test fixtures (`tests/conftest.py`)

- Mock WebSocket connection fixture that returns configurable JSON responses
- Mock `DaemonServer` credentials fixture
- Factory functions for `QueryResult`, `ConnectionResult`, and other server response types
- Fixtures that configure the mock to simulate error conditions

### 2. Create unit tests for core components

Priority modules to test in isolation:

| Module | What to test |
|--------|-------------|
| `Query` / `PoolQuery` | State machine transitions (NOT_YET_RUN -> RUN_MORE_DATA_AVAIL -> RUN_DONE -> ERROR) |
| `Pool` | Scaling logic (add/remove jobs), job selection algorithm, pool exhaustion |
| `DaemonServer` | Config parsing from dicts, INI files, kwargs; validation of required fields |
| `core/exceptions.py` | Error classification mapping (RuntimeError -> PEP 249 exceptions) |
| `core/cursor.py` | Cursor lifecycle, fetchone/fetchall behavior, closed cursor detection |
| `core/connection.py` | Connection state management, cursor creation |
| `base_job.py` | `_parse_connection_input()` for all input types |
| `ssl.py` | Certificate retrieval and SSL context creation |

### 3. Fix commented-out tests

Investigate and fix the 7 commented-out tests:
- `test_pop_job_with_pool_ignore` — JobStatus tracking
- `test_pool_with_no_space_no_ready_job...` — JobStatus tracking
- `test_pool_with_space_but_no_ready_job...` — JobStatus tracking
- `test_freeist_job_is_returned` — JobStatus tracking
- `test_pep249_execute_many` — Duplicate of working test
- `test_prepare_statement_mult_params_seq_tuple` — Tuple parameter support
- `test_prepare_statement_mult_params_seq_tuple_opts` — Tuple parameter support

### 4. Add coverage reporting

- Configure `.coveragerc` or `pyproject.toml` coverage settings
- Add `--cov=mapepire_python --cov-report=term-missing` to default pytest invocation
- Set a coverage floor (e.g., 70%) as a CI gate

## Files to Create/Modify

- **New**: `tests/conftest.py` — Shared fixtures and mock WebSocket
- **New**: `tests/unit/test_query_state.py` — Query state machine tests
- **New**: `tests/unit/test_pool_scaling.py` — Pool logic tests
- **New**: `tests/unit/test_config_parsing.py` — DaemonServer/config tests
- **New**: `tests/unit/test_error_classification.py` — Exception mapping tests
- **New**: `tests/unit/test_cursor.py` — Cursor lifecycle tests
- **Modify**: `tests/pooling_test.py` — Fix commented-out tests
- **Modify**: `tests/pep249_test.py` — Fix commented-out tests
- **Modify**: `pyproject.toml` — Add coverage configuration

## Acceptance Criteria

- [ ] `conftest.py` provides reusable mock WebSocket fixtures
- [ ] At least 15 new unit tests covering the modules listed above
- [ ] Unit tests run without any server connection (`pytest tests/unit/`)
- [ ] All 7 commented-out tests are either fixed or replaced with equivalent coverage
- [ ] Coverage reporting is configured and runs with `pytest --cov`
- [ ] Unit tests pass in CI without IBM i server credentials

### PR DESCRIPTION
**Pull request by Vnajeb**
Fixes #92 

Changes proposed in this pull request:

- Split the test suite into tests/unit/ (offline) and tests/integration/ (live IBM i server). The offline suite mocks the WebSocket layer via shared conftest.py fixtures, so connection, cursor, pool-scaling, query-state, error-classification, config-parsing, and SSL behavior can all be tested without a live server.
- Wire up coverage: add `pytest-cov` config and enable it by default via `addopts`
- Resolve 7 previously commented-out tests and re-enable them in the existing suite
- Split tests into tests/unit/ and tests/integration/. This means unit tests no longer need a server or credentials to run.
- Split CI into two separate jobs: Unit Test (no secrets needed) and Integration Test (needs VITE_* secrets, skipped on forks). Previously these ran together in the same matrix.

## Before submitting

- [X] Change the base branch to `dev` if it is not already.
- [X] I've read and followed all steps in the [Making a pull request](https://github.com/Mapepire-IBMi/mapepire-python/blob/main/.github/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [X] I've updated or added any relevant docstrings, following the Google docstring style used
    elsewhere in the codebase.
- [X] If this PR fixes a bug, I've added a test that will fail without my fix.
- [X] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -125,6 +126,26 @@ log_level = "DEBUG"
...
+[tool.coverage.run]
+source = ["mapepire_python"]
+branch = true
+omit = [
+  "tests/*",
+  "*/__pycache__/*",
+]
+
+[tool.coverage.report]
+fail_under = 70
```
Comment (rousseauc1) [inline pyproject.toml]: As of now this causes the pytest tests/unit/ to fail at around 66% coverage despite having 160 passing tests, so you should remove the global floor and enfore --cov-fail-under=70 on just full/integration CI jobs, or keep it and pass --cov-fail-under=0 on the unit-test jobs.

File: tests/pep249_test.py [hunk 2]
```diff
@@ -110,22 +110,23 @@ def test_prepare_statement_mult_params_seq():
...
-
-
-# def test_prepare_statement_mult_params_seq_tuple_opts():
-#     conn = connect(creds)
-#     cur = conn.cursor()
-#     opts = QueryOptions(parameters=[[500, "PRES"], [200, "PRES"]])
-#     cur.execute("select * from sample.employee where bonus > ? and job = ?", opts=opts)
-#     res = cur.fetchall()
-#     assert res["success"] == True
+# NOTE: A removed test_prepare_statement_mult_params_seq_tuple passed multiple
```
Comment (rousseauc1) [inline tests/pep249_test.py]: Note only mentions the first test, might want to expand here and add a line or two about the second test that was removed the tuple_opts

**Review by rousseauc1 — CHANGES_REQUESTED** [review]
Tests look good besides a few changes. I'd say add in separate testing in .github/workflows/main.yml so that when you run it, it doesn't run both integration and unit tests. Issue #92 wants to be able to run the unit tests offline and because the integration tests are combined in that matrix job it does need secrets (see fix example below). And then move the integration tests into tests/integration/ to keep them organized.



```bash
task:
  - name: Unit Test
    run: pytest -v --color=yes tests/unit/
  - name: Integration Test
    run: pytest -v --color=yes tests/integration/

**Review by rousseauc1 — APPROVED** [review]
